File: backtracking/Lotto.py
import sys
input = sys.stdin.readline

def solve(a,k,num,arr,result):
    if a==6:
        for i in range(6): print(result[i], end=" ")
        print()
        return
    
    st = arr[a-1]+1 if a!=0 else 0
    for i in range(st,k):
        result[a] = num[i]
        arr[a] = i
        solve(a+1,k,num,arr,result)

def main():
    while(1):
        data = list(map(int, input().split()))
        k = data[0]
        if k==0: break
        num = data[1:]
        num = sorted(num)
        arr = [0]*6
        result = [0]*6
        solve(0,k,num,arr,result)
        print()

if __name__ == "__main__":
    main()

File: backtracking/Lotto.cpp
#include<bits/stdc++.h>
using namespace std;

int k;
int num[15];
int arr[10];

void solve(int a, int st){
    if(a==6){
        for(int i=0;i<6;i++) cout<<arr[i]<<" ";
        cout<<"\n";
        return;
    }
    for(int i=st;i<k;i++) {
        arr[a] = num[i];
        solve(a+1,i+1);
    }
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    while(1){
        cin>>k;
        if(k==0) break;
        for(int i=0;i<k;i++) cin>>num[i];
        sort(num,num+k);
        solve(0,0);
        cout<<"\n";
    }
}
